File: js/src/signing.js
"use strict";

const crypto = require("crypto");

function buildStringToSign({ method, path: reqPath, timestamp, accountId }) {
  return `${String(method).toUpperCase()}\n${reqPath}\n${timestamp}\n${accountId}`;
}

function signRequestHeaders({
  accessKeyId,
  secret,
  method,
  path: reqPath,
  accountId,
  sessionToken,
}) {
  const timestamp = new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
  const stringToSign = buildStringToSign({
    method,
    path: reqPath,
    timestamp,
    accountId,
  });
  const signature = crypto.createHmac("sha256", secret).update(stringToSign).digest("hex");
  const headers = {
    "X-Homecloud-Access-Key-Id": accessKeyId,
    "X-Homecloud-Date": timestamp,
    "X-Homecloud-Signature": signature,
  };
  if (sessionToken) headers["X-Homecloud-Session-Token"] = sessionToken;
  return headers;
}

function encodeObjectKeyPath(key) {
  return key
    .replace(/^\/+/, "")
    .split("/")
    .map((part) => encodeURIComponent(part))
    .join("/");
}

function soObjectPaths(accountId, bucketName, objectKey) {
  const key = String(objectKey || "").replace(/^\/+/, "");
  const signPath = `/${accountId}/${bucketName}/objects/${key}`;
  const urlPath = `/${accountId}/${bucketName}/objects/${encodeObjectKeyPath(key)}`;
  return { signPath, urlPath };
}

module.exports = {
  buildStringToSign,
  signRequestHeaders,
  encodeObjectKeyPath,
  soObjectPaths,
};

File: go/signing.go
package homecloud

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"strings"
	"time"
)

const (
	HeaderAccessKeyID  = "X-Homecloud-Access-Key-Id"
	HeaderDate         = "X-Homecloud-Date"
	HeaderSignature    = "X-Homecloud-Signature"
	HeaderSessionToken = "X-Homecloud-Session-Token"
	HeaderIdempotency  = "Idempotency-Key"
)

// BuildStringToSign is the SigV1 canonical string:
//
//	{METHOD}\n{path}\n{timestamp}\n{account_id}
func BuildStringToSign(method, path, timestamp, accountID string) string {
	return strings.ToUpper(method) + "\n" + path + "\n" + timestamp + "\n" + accountID
}

func ComputeSignature(secret, stringToSign string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(stringToSign))
	return hex.EncodeToString(mac.Sum(nil))
}

func FormatTimestamp(t time.Time) string {
	return t.UTC().Truncate(time.Second).Format("2006-01-02T15:04:05Z")
}

func signHeaders(accessKeyID, secret, method, path, accountID string, now time.Time, sessionToken string) map[string]string {
	ts := FormatTimestamp(now)
	sig := ComputeSignature(secret, BuildStringToSign(method, path, ts, accountID))
	h := map[string]string{
		HeaderAccessKeyID: accessKeyID,
		HeaderDate:        ts,
		HeaderSignature:   sig,
	}
	if strings.TrimSpace(sessionToken) != "" {
		h[HeaderSessionToken] = sessionToken
	}
	return h
}
